server/controllers/DatabaseController.js

```diff
@@ -0,0 +1,242 @@
+    } else {
+      res.write(`Dropping tables: ${tableNames.join(', ')}\n`);
+      await client.query(`DROP TABLE ${tableNames.join(', ')} CASCADE;`);
+      res.write('All tables have been dropped successfully.\n');
+    }
+    // [Tables DROP // END]
+
+    // [Functions DROP // START]
+    // Get all user-defined functions
+    const functionsQuery = `
+      SELECT 'DROP FUNCTION IF EXISTS ' || n.nspname || '.' || p.proname || '(' ||
+             pg_get_function_identity_arguments(p.oid) || ');' AS drop_statement
+      FROM pg_proc p
+             JOIN pg_namespace n ON p.pronamespace = n.oid
+      WHERE n.nspname NOT IN ('pg_catalog', 'information_schema');
+    `;
+
+    const functionsResult = await client.query(functionsQuery);
+    const dropStatements = functionsResult.rows.map(row => row.drop_statement);
+
+    // Execute each DROP FUNCTION statement
+    for (const dropStatement of dropStatements) {
+      res.write(`Executing: ${dropStatement}\n`);
+      await client.query(dropStatement);
+    }
+    res.write('All functions removed successfully!\n');
+    // [Functions DROP // END]
+  } catch (error) {
+    res.write(
+      `Error removing tables or functions: ${sanitizeError(error.message)}\n`
+    );
+  } finally {
+    await client.end();
+    res.write('Disconnected from the database.\n');
+  }
+};
+
+const dumpPostgresDatabase = async (req, res) => {
+  const { dumpOutputDirectory, dumpFileName } = req.body;
+
+  if (dumpFileName && !dumpFileName.includes('.sql'))
+    return res.status(400).send("Provided file name does not include '.sql'");
+
+  const OUTPUT_DIR = dumpOutputDirectory || path.join(os.homedir(), 'db_dumps');
+  const DUMP_FILE = path.join(
+    OUTPUT_DIR,
+    dumpFileName ||
+      `${process.env.ENVIRONMENT}_dump_${dates.convertDateToString(
+        new Date(),
+        'YYYYMMDD_HHmmss'
+      )}.sql`
+  );
+
+  try {
+    // pg_dump command
+    const pgDumpCommand = `PGPASSWORD=${DB_PASSWORD} pg_dump -U ${DB_USER} -h ${DB_HOST} -d ${DB_NAME} --exclude-table=session > ${DUMP_FILE}`;
+
+    // Execute the command
+    exec(pgDumpCommand, (error, stdout, stderr) => {
+      if (error) {
+        return res
+          .status(400)
+          .send(`Error executing pg_dump: ${sanitizeError(error.message)}`);
+      }
+      if (stderr) {
+        return res
+          .status(400)
+          .send(`pg_dump stderr:: ${sanitizeError(stderr)}`);
+      }
+      return res
+        .status(200)
+        .send(`Database dump completed successfully: ${DUMP_FILE}`);
+    });
+  } catch (error) {
+    return res
+      .status(400)
+      .send(`Unable to dump database: ${sanitizeError(error.message)}`);
+  }
+};
+
+const restorePostgresDatabase = async (req, res) => {
+  // Prevent unintentionally dropping or restoring database tables if ran on
+  // production unless manually done
+  if (process.env.ENVIRONMENT === 'production') {
+    return res.status(405).send('This request is not permitted');
+  }
```

Comment:
I know, intellectually, that this is probably enough to prevent this being run on prod but the string matching as the only safeguard leaves me a bit uneasy

Reply:
Makes sense! Can also confirm the host as well